File: tests/sdk/go/ec2_nested_test.go
package sdk_test

import (
	"context"
	"strings"
	"testing"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/ec2"
	"github.com/aws/aws-sdk-go-v2/service/ec2/types"
)

func TestEC2RunStopStartTerminatePendingOK(t *testing.T) {
	requireReady(t)
	cfg := loadAWSConfig(t)
	c := ec2.NewFromConfig(cfg, func(o *ec2.Options) {
		o.BaseEndpoint = aws.String(endpoint())
	})
	ctx := context.Background()

	run, err := c.RunInstances(ctx, &ec2.RunInstancesInput{
		ImageId:      aws.String("ami-alpine"),
		InstanceType: types.InstanceTypeT3Micro,
		MinCount:     aws.Int32(1),
		MaxCount:     aws.Int32(1),
	})
	if err != nil {
		t.Fatalf("RunInstances: %v", err)
	}
	if len(run.Instances) != 1 || run.Instances[0].InstanceId == nil {
		t.Fatalf("instances=%+v", run.Instances)
	}
	id := aws.ToString(run.Instances[0].InstanceId)
	if !strings.HasPrefix(id, "i-") {
		t.Fatalf("instance id=%q", id)
	}
	t.Cleanup(func() {
		_, _ = c.TerminateInstances(ctx, &ec2.TerminateInstancesInput{
			InstanceIds: []string{id},
		})
	})

	desc, err := c.DescribeInstances(ctx, &ec2.DescribeInstancesInput{
		InstanceIds: []string{id},
	})
	if err != nil {
		t.Fatalf("DescribeInstances: %v", err)
	}
	if len(desc.Reservations) == 0 || len(desc.Reservations[0].Instances) == 0 {
		t.Fatal("DescribeInstances empty")
	}
	state := ""
	if desc.Reservations[0].Instances[0].State != nil {
		state = string(desc.Reservations[0].Instances[0].State.Name)
	}
	if state != "pending" && state != "running" {
		t.Fatalf("after Run want pending|running got %q", state)
	}

	stop, err := c.StopInstances(ctx, &ec2.StopInstancesInput{InstanceIds: []string{id}})
	if err != nil {
		t.Fatalf("StopInstances: %v", err)
	}
	if len(stop.StoppingInstances) < 1 {
		t.Fatal("expected StoppingInstances")
	}

	start, err := c.StartInstances(ctx, &ec2.StartInstancesInput{InstanceIds: []string{id}})
	if err != nil {
		t.Fatalf("StartInstances: %v", err)
	}
	if len(start.StartingInstances) < 1 {
		t.Fatal("expected StartingInstances")
	}
	// Without noctaxris-engine, Start leaves pending; with engine, running.
	started := ""
	if start.StartingInstances[0].CurrentState != nil {
		started = string(start.StartingInstances[0].CurrentState.Name)
	}
	if started != "pending" && started != "running" {
		t.Fatalf("after Start want pending|running got %q", started)
	}

	term, err := c.TerminateInstances(ctx, &ec2.TerminateInstancesInput{InstanceIds: []string{id}})
	if err != nil {
		t.Fatalf("TerminateInstances: %v", err)
	}
	if len(term.TerminatingInstances) < 1 {
		t.Fatal("expected TerminatingInstances")
	}
}

File: tests/sdk/nodejs/test/ec2-nested.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  DescribeInstancesCommand,
  EC2Client,
  RunInstancesCommand,
  StartInstancesCommand,
  StopInstancesCommand,
  TerminateInstancesCommand,
} from "@aws-sdk/client-ec2";
import { endpoint, region, requireReady } from "../lib/helpers.mjs";

function baseClientConfig() {
  return {
    region: region(),
    endpoint: endpoint(),
    credentials: {
      accessKeyId: process.env.AWS_ACCESS_KEY_ID || "AKIAROOTEXAMPLE01",
      secretAccessKey:
        process.env.AWS_SECRET_ACCESS_KEY ||
        "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY",
    },
  };
}

test("EC2 Run Stop Start Terminate pending OK", async (t) => {
  if (!(await requireReady(t))) return;
  const c = new EC2Client(baseClientConfig());

  const run = await c.send(
    new RunInstancesCommand({
      ImageId: "ami-alpine",
      InstanceType: "t3.micro",
      MinCount: 1,
      MaxCount: 1,
    }),
  );
  const id = run.Instances?.[0]?.InstanceId;
  assert.ok(id, `missing InstanceId: ${JSON.stringify(run)}`);
  assert.match(id, /^i-/);

  try {
    const desc = await c.send(
      new DescribeInstancesCommand({ InstanceIds: [id] }),
    );
    const state = desc.Reservations?.[0]?.Instances?.[0]?.State?.Name || "";
    assert.ok(
      state === "pending" || state === "running",
      `after Run want pending|running got ${state}`,
    );

    const stop = await c.send(new StopInstancesCommand({ InstanceIds: [id] }));
    assert.ok((stop.StoppingInstances || []).length >= 1);

    const start = await c.send(new StartInstancesCommand({ InstanceIds: [id] }));
    assert.ok((start.StartingInstances || []).length >= 1);
    const started = start.StartingInstances?.[0]?.CurrentState?.Name || "";
    assert.ok(
      started === "pending" || started === "running",
      `after Start want pending|running got ${started}`,
    );

    const term = await c.send(
      new TerminateInstancesCommand({ InstanceIds: [id] }),
    );
    assert.ok((term.TerminatingInstances || []).length >= 1);
  } finally {
    try {
      await c.send(new TerminateInstancesCommand({ InstanceIds: [id] }));
    } catch {
      /* already terminated */
    }
  }
});
